src/Intervals.php

```diff
@@ -0,0 +1,369 @@
+     *
+     * @return array
+     * @phpstan-return array{'intervals': array<int, array{'start': Constraint, 'end': Constraint}>, 'devConstraints': Constraint[]}
+     */
+    public static function get(ConstraintInterface $constraint)
+    {
+        $key = (string) $constraint;
+
+        if (!isset(self::$intervalsCache[$key])) {
+            self::$intervalsCache[$key] = self::generateIntervals($constraint);
+        }
+
+        return self::$intervalsCache[$key];
+    }
+
+    /**
+     * @phpstan-return array{'intervals': array<int, array{'start': Constraint, 'end': Constraint}>, 'devConstraints': Constraint[]}
+     */
+    private static function generateIntervals(ConstraintInterface $constraint, $stopOnFirstValidInterval = false)
+    {
+        if ($constraint instanceof EmptyConstraint) {
+            return array('intervals' => array(array('start' => self::zero(), 'end' => self::positiveInfinity())), 'devConstraints' => array(self::anyDev()));
+        }
+
+        if ($constraint instanceof Constraint) {
+            return self::generateSingleConstraintIntervals($constraint);
+        }
+
+        if (!$constraint instanceof MultiConstraint) {
+            throw new \UnexpectedValueException('The constraint passed in should be an EmptyConstraint, Constraint or MultiConstraint instance, got '.get_class($constraint).'.');
+        }
+
+        $constraints = $constraint->getConstraints();
+
+        $intervalGroups = array();
+        $devGroups = array();
+        foreach ($constraints as $c) {
+            $res = self::get($c);
+            $intervalGroups[] = $res['intervals'];
+            $devGroups[] = $res['devConstraints'];
+        }
+
+        $dev = array();
+        if ($constraint->isDisjunctive()) {
+            foreach ($devGroups as $group) {
+                foreach ($group as $c) {
+                    if (!isset($dev[(string) $c])) {
+                        $dev[(string) $c] = $c;
+                    }
+                }
+            }
+
+            foreach ($dev as $i => $c) {
+                if ($c->getOperator() === '!=') {
+                    foreach ($dev as $j => $c2) {
+                        if ($i === $j) {
+                            continue;
+                        }
+                        $op = $c2->getOperator();
+                        // != dev-foo || != dev-bar -> *
+                        if ($op === '!=' && $c->getVersion() !== $c2->getVersion()) {
+                            $dev = array();
+                            break 2;
+                        }
+                        // != dev-foo || == dev-foo -> *
+                        if ($op === '==' && $c->getVersion() === $c2->getVersion()) {
+                            $dev = array();
+                            break 2;
+                        }
+                        // != dev-foo || == dev-master -> != dev-foo
+                        if ($op === '==') {
```

Comment:
this behaves incorrectly for anyDev, e.g. `!= dev-foo || == dev*` -> `*` not `!= dev-foo`

Reply:
Should probably also add a test that covers this situation.

Reply:
Fixed in https://github.com/composer/semver/pull/97/commits/3d72ad3c44db65b986e54f0a9ff818beb8c6e066